### PR TITLE
Upstream fixes

### PR DESCRIPTION
**Pull request by thomaslitton**
## Pull Request Checklist

**Is this in reference to an existing issue?**

#### General

- [ ] Update Changelog following the conventions laid out [here](https://github.com/sensu-plugins/community/blob/master/HOW_WE_CHANGELOG.md)

- [x] Update README with any necessary configuration snippets

- [x] Binstubs are created if needed

- [x] RuboCop passes

- [x] Existing tests pass

#### New Plugins

- [x] Tests

- [x] Add the plugin to the README

- [x] Does it have a complete header as outlined [here](http://sensu-plugins.io/docs/developer_guidelines.html#coding-style)

### CODE REVIEW

**Comment by majormoses** [comment]
Thanks for your contribution to Sensu plugins! Without people like you submitting PRs we couldn't run the project. I will review it shortly.

File: CHANGELOG.md
Comment (majormoses) [inline CHANGELOG.md]: per our [changelog conventions](https://github.com/sensu-plugins/community/blob/master/HOW_WE_CHANGELOG.md) maintainers version and date the releases post acceptance. There are several reasons for this:
- PR review and merge is not FIFO
- there is no SLA on on times waiting for review, I try to review within 24-48 hours. While I often review much quicker than that there are times where it takes me several days to get to them
- maintainer might disagree on the impact of a particular change
- As PRs do not necessarily get merged in the order they receive not including the version reduces the number of times we have to ask for a rebase or updating the value.

File: bin/check-azurerm-monitor-metric.rb
Comment (majormoses) [inline bin/check-azurerm-monitor-metric.rb]: based on the error message it should be `||` not `&&` update the logic or message to indicate what is right.

File: CHANGELOG.md
Comment (majormoses) [inline CHANGELOG.md]: Ideally a little more info will help people a monitoring check is quite ambiguous. Maybe something like this:
```
- check-azurerm-monitor-metric.rb: allows you to check against azure metric thresholds for a particular resource id or name (@thomaslitton)
```

File: bin/check-azurerm-monitor-metric.rb
Comment (majormoses) [inline bin/check-azurerm-monitor-metric.rb]: why is this a string? It probably should be an integer right?

File: bin/check-azurerm-monitor-metric.rb
Comment (majormoses) [reply bin/check-azurerm-monitor-metric.rb]: Given bellow this implies the logic should change

File: bin/check-azurerm-monitor-metric.rb
Comment (majormoses) [inline bin/check-azurerm-monitor-metric.rb]: So you could not say specify a utilization range then? I think there are use cases where you want things to stay in a range. Queued worker messages for backend batch or async processing work comes to mind. Network utilization from a NAT comes to mind, if there is ever under a certain amount of traffic there is something horribly wrong if its too high you need to look into scale your NAT throughput.

I am not asking you to add it just talking through some things that someone might want in the future. You might find you have use cases you did not think about and want this even if its in another PR.

If we wanted to do this we could just keep track of the critical, warnings for each and at the end we could include all crit and/or warning

File: bin/check-azurerm-monitor-metric.rb
Comment (majormoses) [inline bin/check-azurerm-monitor-metric.rb]: why cast it to a symbol and then a float?

File: bin/check-azurerm-monitor-metric.rb
Comment (majormoses) [inline bin/check-azurerm-monitor-metric.rb]: It's been my experience with azure that they have a bunch of different portals and they change them regularly perhaps we should expose this so if they ever change it people can update their runtime options and unblock themselves without waiting for a fixed release. Not a requirement but just a thought.

File: bin/check-azurerm-monitor-metric.rb
Comment (majormoses) [inline bin/check-azurerm-monitor-metric.rb]: seeing as we have seen this url twice in the file we should at least variablize it if not making it a cli arg.

File: bin/check-azurerm-monitor-metric.rb
Comment (majormoses) [inline bin/check-azurerm-monitor-metric.rb]: I think that the timespan should be configurable, I can't say I know much about how azure does metrics but I do look at cloudwatch metrics in aws in shorter intervals than this for some metrics.

File: lib/sensu-plugins-azurerm/version.rb
Comment (majormoses) [inline lib/sensu-plugins-azurerm/version.rb]: see previous comment about maintainers bumping versions post acceptance.

**Review by majormoses — CHANGES_REQUESTED** [review]
This looks really great, I have some small suggestions to make. I really appreciate the tests.

File: bin/check-azurerm-monitor-metric.rb
Comment (thomaslitton) [reply bin/check-azurerm-monitor-metric.rb]: If it were an or, that would require the user to provide both correct?  Only one is needed.

File: bin/check-azurerm-monitor-metric.rb
Comment (thomaslitton) [reply bin/check-azurerm-monitor-metric.rb]: It is not.  It is casting the option used as the key to a symbol, and the value from the hash to a float.

Comment (majormoses) [reply bin/check-azurerm-monitor-metric.rb]: you are right, this is what happens when I spend too much time reviewing PRs.

Comment (majormoses) [reply bin/check-azurerm-monitor-metric.rb]: you could do a conditional assignment above to reduce code duplication but by no means a requirement.

File: bin/check-azurerm-monitor-metric.rb
Comment (majormoses) [reply bin/check-azurerm-monitor-metric.rb]: ah I was thrown off by the verbiage in the message. I will comment below on what I think we can do to make it clearer.

File: bin/check-azurerm-monitor-metric.rb
Comment (majormoses) [inline bin/check-azurerm-monitor-metric.rb]: Given the above I think we can improve this, something along the lines of:
```
unknown "resource id, resource name/group/type/namespace, and subscription id are required. You passed: resource_id: #{config[:resource_id]}, resource_name: #{config[:resource_name]}"
```

Also the error message indicates that subscription id must be specified but we do not check that, we can just make it a required parameter. I think if we make resource id a parameter we should check if any of `$resource_( name|group|namespace)^` are valid rather than just checking 2 possible options. Again I am going off the old message so if resource_name is required we should just make it required but if one of any of those are required we can check that. Also if you want you can remove the default strings and check if they are `nil` instead. You can also do a `proc` to to the `.to_s` so you don't have to do that here as well every time you use it.

File: bin/check-azurerm-monitor-metric.rb
Comment (majormoses) [inline bin/check-azurerm-monitor-metric.rb]: other than the trailing `/` this looks the same as `base_url` would they ever be different? If not lets use one option and where we need the slash it can be part of the implementation itself. Maybe this can be used against an internal AD or something but I don't know much about AD and Azure to be honest.

**Review by majormoses — CHANGES_REQUESTED** [review]

File: bin/check-azurerm-monitor-metric.rb
Comment (thomaslitton) [reply bin/check-azurerm-monitor-metric.rb]: The resource id is made up of resource name/group/type/namespace and subscription id (along an optional resource parent).  Therefore, if you specify the resource id, the rest isn't used.  I'll make changes to make this clearer.

File: bin/check-azurerm-monitor-metric.rb
Comment (thomaslitton) [reply bin/check-azurerm-monitor-metric.rb]: I don't know if they will always be the same.  Plus one is used all the time, and one is used some of the time, so i thought it would be clearer to be 2 separate options.

Comment (majormoses) [reply bin/check-azurerm-monitor-metric.rb]: ok just wanted to see if there was a valid reason and if so I am :+1:

File: bin/check-azurerm-monitor-metric.rb
Comment (thomaslitton) [reply bin/check-azurerm-monitor-metric.rb]: I'll try to fit this change in.

File: README.md
Comment (tombuildsstuff) [inline README.md:436]: would it be worth making this example more complete via an example such as `/subscriptions/0000../resourceGroups/foo/...`?

File: bin/check-azurerm-monitor-metric.rb
Comment (tombuildsstuff) [inline bin/check-azurerm-monitor-metric.rb:293]: not needed for this PR - but longer term we could probably split this out into a helper function?

**Review by tombuildsstuff — APPROVED** [review]
hey @thomaslitton

Thanks for this PR - I've taken a look through and I've left one minor comment (in addition to @majormoses comments) - but this otherwise LGTM 👍

Thanks!

**Comment by thomaslitton** [comment]
@majormoses  Are there any more changes that are needed for this?

**Comment by majormoses** [comment]
Sorry this fell off my radar, looke like there are some questions in the review comments still unaddressed can you address them?

**Review by majormoses — APPROVED** [review]
LGTM

**Comment by majormoses** [comment]
released: https://rubygems.org/gems/sensu-plugins-azurerm/versions/2.1.0

**Pull request merged**
